week4/solutions/scrabble/tests/test_hand.py

```diff
@@ -0,0 +1,480 @@
+    vowels_test(self, 11)
+
+
+# Hand.get_consonants()
+CONSONANT_SET = set(['b', 'c', 'd', 'f', 'g', 'h', 'j', 'k', 'l', 'm', 'n',
+                  'p', 'q', 'r', 's', 't', 'v', 'w', 'x', 'y', 'z'])
+
+def consonant_test(self, hand_size):
+  hand = Hand(hand_size)
+  hand.get_consonants()
+  number_consonants = sum(hand.get().values())
+  expected_number_consonants = hand_size - hand_size//3
+  are_consonants = set(hand.get().keys()) <= CONSONANT_SET
+  
+  self.assertEqual(number_consonants, expected_number_consonants)
+  self.assertEqual(are_consonants, True)
+
+class HandGetConsonantsTest1(unittest.TestCase):
+  def test(self):
+    '''hand = Hand(30)
+    hand.get_consonants()
+    number_consonants = sum(hand.get()[key] for key in hand.get())
+    are_consonants = set(hand.get().keys()) <= CONSONANT_SET
+    self.assertEqual(number_consonants, 20)
+    self.assertEqual((set(hand.get().keys()) <= CONSONANT_SET), True)'''
+    consonant_test(self, 30)
+
+class HandGetConsonantsTest2(unittest.TestCase):
+  def test(self):
+    '''hand = Hand(5)
+    hand.get_consonants()
+    number_consonants = sum(hand.get()[key] for key in hand.get())
+    are_consonants = set(hand.get().keys()) <= CONSONANT_SET
+    self.assertEqual(number_consonants, 4)
+    self.assertEqual((set(hand.get().keys()) <= CONSONANT_SET), True)'''
+    consonant_test(self, 5)
+
+class HandGetConsonantsTest3(unittest.TestCase):
+  def test(self):
+    '''hand = Hand(9)
+    hand.get_consonants()
+    number_consonants = sum(hand.get()[key] for key in hand.get())
+    are_consonants = set(hand.get().keys()) <= CONSONANT_SET
+    self.assertEqual(number_consonants, 6)
+    self.assertEqual((set(hand.get().keys()) <= CONSONANT_SET), True)'''
+    consonant_test(self, 9)
+
+class HandGetConsonantsTest4(unittest.TestCase):
+  def test(self):
+    '''hand = Hand(11)
+    hand.get_consonants()
+    number_consonants = sum(hand.get()[key] for key in hand.get())
+    are_consonants = set(hand.get().keys()) <= CONSONANT_SET
+    self.assertEqual(number_consonants, 8)
+    self.assertEqual((set(hand.get().keys()) <= CONSONANT_SET), True)'''
+    consonant_test(self, 11)
+
+
+# Hand.calculate_len()
+class CalculateHandLenTest1(unittest.TestCase):
+  def test(self):
+    hand = Hand()
+    hand.set({'e':1, 'v':2, 'n':1, 'i':1, 'l':2})
+    
+    self.assertEqual(hand.calculate_len(), 7)
+
+class CalculateHandLenTest2(unittest.TestCase):
+  def test(self):
+    hand = Hand(15)
+    hand.set({'e':5, 'v':2, 'n':4, 'i':2, 'l':2})
+    
+    self.assertEqual(hand.calculate_len(), 15)
+
+class CalculateHandLenTest3(unittest.TestCase):
+  def test(self):
+    hand = Hand()
+    hand.set({'e':1, 'v':-2, 'n':1, 'i':-1, 'l':2})
+    self.assertEqual(hand.calculate_len(), 1)
+
+class CalculateHandLenTest4(unittest.TestCase):
+  def test(self):
+    hand = Hand()
+    self.assertEqual(hand.calculate_len(), 0)
+
+
+# Hand.play()
+# need to figure how to mock the game play
```

Comment:
I would like to be able to mock the full game play, but I not quite figured it out. I am also not confident that I am using effectively in general. I think I need to figure out when it is good to mock and when it is not. The way I have applied it in test_hand.py and test_game.py feels pretty clumsy to me.

Reply:
I see what you're saying. Ideally, you'll have one or two test classes per module. Inside there you'll have a series of methods that unit test your code. This is an example of that: https://github.com/VolusionDev/affiliation-attribution-service/blob/master/tests/app_api_test.py. See how clean that looks? Without actually understanding the code, it just looks clean. This is usually (but not always) a good sign.

So, you'd refactor your code into something like this:

```python
class HandGetTest1(unittest.TestCase):
  def setUp(self):  # I think this is the right name
      hand = Hand()

  def test_empty_hand(self):
    self.assertEqual(isinstance(self.hand.get(), dict), True)
    self.assertEqual(len(self.hand.get()), 0)
    self.assertEqual(self.hand.get(), {})

  def test_hand6(self):
    hand = self.hand(6)
    self.assertEqual(isinstance(hand.get(), dict), True)
    self.assertEqual(len(hand.get()), 0)
    self.assertEqual(hand.get(), {})

  def test_deal(self):
    self.hand.deal()
    letters_in_hand = sum(hand.get().values())
    print(hand.get())
```
...etc.